### PR TITLE
Add an endpoint to download a file

### PR DESCRIPTION
**Pull request by a3957273**

### CODE REVIEW

File: backend/src/clients/s3.ts [hunk 1]
```diff
@@ -31,3 +31,21 @@ export async function putObjectStream(bucket: string, key: string, body: Readabl
     fileSize,
   }
 }
+
+export async function getObjectStream(bucket: string, key: string, range?: { start: number; end: number }) {
```
Comment (JR40159) [inline backend/src/clients/s3.ts:35]: Tests for this method?

File: backend/src/clients/s3.ts [hunk 2]
```diff
@@ -31,3 +31,21 @@ export async function putObjectStream(bucket: string, key: string, body: Readabl
...
+  const client = await getS3Client()
```
Comment (JR40159) [inline backend/src/clients/s3.ts:36]: Do we need to create a new S3 Client for every S3 request?

File: backend/src/routes/v2/model/file/getDownloadFile.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+import { FileInterface } from '../../../../models/v2/File.js'
+import { downloadFile, getFileById } from '../../../../services/v2/file.js'
+import { BadReq, InternalError } from '../../../../utils/v2/error.js'
+import { parse } from '../../../../utils/validate.js'
+
+export const getDownloadFileSchema = z.object({
+  params: z.object({
+    modelId: z.string({
+      required_error: 'Must specify model id as param',
```
Comment (JR40159) [inline backend/src/routes/v2/model/file/getDownloadFile.ts:14]: The standard Zod error here would be sufficient?

File: backend/src/routes/v2/model/file/getDownloadFile.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    } = parse(req, getDownloadFileSchema)
+
+    const file = await getFileById(req.user, fileId)
+
+    // required to support utf-8 file names
+    res.set('Content-Disposition', contentDisposition(file.name, { type: 'inline' }))
+    res.set('Content-Type', file.mime)
+    res.set('Cache-Control', 'public, max-age=604800, immutable')
+
+    if (req.headers.range) {
```
Comment (JR40159) [inline backend/src/routes/v2/model/file/getDownloadFile.ts:38]: Do this first before logic for response?

File: backend/src/routes/v2/model/file/getDownloadFile.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    // res.set('Accept-Ranges', 'bytes')
+    res.writeHead(200)
+
+    const stream = await downloadFile(req.user, fileId)
+    if (!stream.Body) {
+      throw InternalError('We were not able to retrieve the body of this file', { fileId })
+    }
+
+    // The AWS library doesn't seem to properly type 'Body' as being pipeable?
+    ;(stream.Body as any).pipe(res)
```
Comment (JR40159) [inline backend/src/routes/v2/model/file/getDownloadFile.ts:54]: Can you raise this issue on their repo?

File: backend/src/routes/v2/model/file/getDownloadFile.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  },
```
Comment (JR40159) [inline backend/src/routes/v2/model/file/getDownloadFile.ts:55]: Am I missing something, where is the method's return statement?

File: backend/src/services/v2/file.ts [hunk 7]
```diff
@@ -37,6 +37,18 @@ export async function uploadFile(user: UserDoc, modelId: string, name: string, m
   return file
 }
 
+export async function downloadFile(user: UserDoc, fileId: string, range?: { start: number; end: number }) {
```
Comment (JR40159) [inline backend/src/services/v2/file.ts:40]: Tests?

**Review by JR40159 — CHANGES_REQUESTED** [review]

File: backend/src/clients/s3.ts (hunk 2 above)
Comment (a3957273) [reply backend/src/clients/s3.ts:36]: Our S3 credentials expire on some platforms relatively frequently.  Tracking whether the current S3 client has expired and replacing it seems too complex for a function that's likely to only get called a few hundred - thousand times a day.

File: backend/src/routes/v2/model/file/getDownloadFile.ts (hunk 4 above)
Comment (a3957273) [reply backend/src/routes/v2/model/file/getDownloadFile.ts:38]: Range requests will include the same headers (Content-Type, Cache-Control and Content-Disposition) seen above.  The idea behind placing it here is that when we _do_ support it you can just write the implementation in here and not need to change anything else.

I could move the error up (rule: do error checking first), but it feels like a waste for now.

**Pull request merged**